### DISCUSSION TITLE
What *is* the difference between "signature information" (`signature_help()`) and "hover information" (`hover()`)?

### DISCUSSION THREAD

**Discussion by pepper-lees** [neovim/neovim#25711] [Q&A] [answered]
Hi,

I'm pretty confused about the two methods, `vim.lsp.buf.hover()` and `vim.lsp.buf.signature_help()`, what their differences are, and what the need is for two separate ones.

If I have the following Scala/Chisel code:

```scala
    val foo = Wire(Bool())
    val bar = Wire(UInt(1.W))

    bar := Mux(foo, 1.U, 0.U)
```

---

Cursor over one of the letters in `bar` on the last line:

`hover()` shows this:
![image](https://github.com/neovim/neovim/assets/11487870/040cb8e7-9891-493f-99f9-33b56e11eaed)
and `signature_help()` does nothing.

---

Cursor over the `:=`:

`hover()` shows this:
![image](https://github.com/neovim/neovim/assets/11487870/2e41deb9-beae-47d4-80d7-1662082d8ed7)

and `signature_help()` does nothing.

---

Cursor over the `Mux`:

`hover()` shows this:
![image](https://github.com/neovim/neovim/assets/11487870/dc08cd2c-f1f7-4fac-a70a-89013ac37e49)

and `signature_help()` shows this - Almost the same as `hover()` when on top of the previous symbol, but not exactly:
![image](https://github.com/neovim/neovim/assets/11487870/663b1176-29c5-45a4-b509-a5fe7037886d)

---

Cursor over the first parenthesis:

`hover()` shows this:
![image](https://github.com/neovim/neovim/assets/11487870/49bdde19-a795-4ad0-b2ed-9e467af4509e)

and `signature_help()` shows this:
![image](https://github.com/neovim/neovim/assets/11487870/dec8960d-5926-4530-a101-fd29d96bb1af)

---

Cursor over `foo`:

`hover()` shows this:
![image](https://github.com/neovim/neovim/assets/11487870/cb7563a5-4fb3-4f52-9757-83ac17208793)

and `signature_help()` shows this:
![image](https://github.com/neovim/neovim/assets/11487870/1a340098-188e-4cfa-97d0-4d54acd3e9e5)

---

Cursor over the `U` in one of the `.U`s:

`hover()` shows this:
![image](https://github.com/neovim/neovim/assets/11487870/cf2d1c6e-59c8-42b9-b5db-0774aa886b93)

and `signature_help()` shows this:
![image](https://github.com/neovim/neovim/assets/11487870/c4cb5be1-9420-49f1-a632-c24679e45232)

---

And lastly, if the cursor is over one of the spaces between the arguments:

`hover()` shows nothing

and `signature_help()` shows this:
![image](https://github.com/neovim/neovim/assets/11487870/a5284386-940e-4027-ab84-20b1043e53be)

---

There's seemingly no pattern to it, except that `hover()` is more useful in almost all cases. More than anything, it's leaving me really confused, for example if `hover()` is the only one giving useful information*, then why does `signature_help()` exist at all? And what _is_ the pattern to it? What is the actual difference between these two functions, and why are there two functions instead of just one?

_*In all cases, the response from `hover()` is the most useful and logical to me, and honestly in the last case, even the fact that `hover()` showed nothing makes the most sense to me, because I was not hovering over any symbols._

Thanks!

**Comment by clason** [comment, marked as answer]
This is explained in the [LSP specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification). Beyond that, it's up to the server what they show exactly, and different servers can and will have (vastly) different contents.

**Reply by pepper-lees** [reply]
Thank you, that document helps a bit. It's helpful that Signature Help has a specific definition (`Signature help represents the signature of something callable. There can be multiple signature but only one active and only one active parameter.`) but i wish that Hover did too. Oh well, I do understand the difference a bit better now.
